runner/src/internal.rs

```diff
@@ -51,15 +51,90 @@ impl Context {
     }
 }
 
+/// The outcome of an individual step of execution.
+#[derive(PartialEq, Eq)]
+pub enum StatusResultValue {
+    Ok,
+    Error,
+    Skipped,
+}
+
+pub struct SingleStatusResult {
+    value: StatusResultValue,
+    command: String,
+    logs: String,
+}
+
+pub struct MultipleStatusResult {
+    statuses: Vec<Status>,
+}
+
+/// The result of a step of execution which may itself be single or multiple.
+pub enum StatusResult {
+    Single(SingleStatusResult),
+    Multiple(MultipleStatusResult),
+}
+
+/// The result of executing a `Runnable` step.
+pub struct Status {
+    name: String,
+    status: StatusResult,
+}
+
+/// A runnable task, which may be implemented as a single or multiple steps.
 pub trait Runnable {
-    fn run(&self, context: &Context);
+    fn run(&self, opt: &Opt) -> Status;
 }
 
 /// Type alias for a boxed, dynamically dispatched `Runnable` instance.
 pub type R = Box<dyn Runnable>;
 
-pub fn run(runnable: R, opt: Opt) {
-    runnable.run(&Context::root(opt))
+/// Run the provided runnable, and print out a summary of the final status (including subtasks).
+pub fn run_all(runnable: R, opt: &Opt) {
+    let root = Context::root(&opt);
+    let status = runnable.run(&opt);
+    print_status(&root, &status);
+}
+
+/// Print the provided status to stderr, applying colors to output when possible.
+fn print_status(context: &Context, status: &Status) {
+    let context = context.child(&status.name);
+    match &status.status {
+        StatusResult::Single(SingleStatusResult {
+            value,
+            logs,
+            command,
+        }) => {
+            match value {
+                StatusResultValue::Ok => {
+                    eprintln!("{} ⊢ {}", context.prefix, "OK".bold().bright_green())
+                }
+                StatusResultValue::Error => {
+                    eprintln!("{} ⊢ {}", context.prefix, "ERROR".bold().bright_red());
+                }
+                StatusResultValue::Skipped => {
+                    eprintln!("{} ⊢ {}", context.prefix, "SKIPPED".bold().bright_yellow())
+                }
+            };
+            if context.opt.commands || context.opt.dry_run {
+                eprintln!("{} ⊢ [{}]", context.prefix, command);
+            };
+            if value == &StatusResultValue::Error || context.opt.logs {
+                eprintln!("{} {}", context.prefix, "╔════════════════════════".blue());
```

Comment:
Does this table need `╗` ?

Reply:
It's only a "half" table, to separate logs from different command, since I don't know how long the log lines actually are.